### PR TITLE
Cuda array interface

### PR DESCRIPTION
**Pull request by LucasGandel**
Implement changes required on the RTK side following the implementation of __cuda_array_interface__ in CudaCommon

Requires [ITKCudaCommon#42](https://github.com/RTKConsortium/ITKCudaCommon/pull/42)

### CODE REVIEW

File: src/rtkCudaWarpForwardProjectionImageFilter.cxx [hunk 1]
```diff
@@ -190,10 +190,10 @@ CudaWarpForwardProjectionImageFilter ::GPUGenerateData()
   inputDVFSize[1] = this->GetDisplacementField()->GetBufferedRegion().GetSize()[1];
   inputDVFSize[2] = this->GetDisplacementField()->GetBufferedRegion().GetSize()[2];
 
-  float * pin = *(float **)(this->GetInputProjectionStack()->GetCudaDataManager()->GetGPUBufferPointer());
-  float * pout = *(float **)(this->GetOutput()->GetCudaDataManager()->GetGPUBufferPointer());
-  float * pvol = *(float **)(this->GetInputVolume()->GetCudaDataManager()->GetGPUBufferPointer());
-  float * pDVF = *(float **)(this->GetDisplacementField()->GetCudaDataManager()->GetGPUBufferPointer());
+  float * pin = (float **)(this->GetInputProjectionStack()->GetCudaDataManager()->etGPUBufferPointer());
```
Comment (SimonRit) [inline src/rtkCudaWarpForwardProjectionImageFilter.cxx]: Typo here

**Review by SimonRit — COMMENTED** [review]
I would like to be able to compile both the CudaCommon latest release and master's head versions, following RTK's policy for ITK versions. In this case, it might be doable by testing whether `GetGPUBufferPointerPtr` is defined as is done in eb90fb42b9dca3fe8be373e2b7c9724e4c3b6156 for an ITK macro. Note the regular expression in the log. Do you want to give it a try or should I try to implement this?

**Comment by LucasGandel** [comment]
> I would like to be able to compile both the CudaCommon latest release and master's head versions, following RTK's policy for ITK versions. In this case, it might be doable by testing whether `GetGPUBufferPointerPtr` is defined as is done in [eb90fb4](https://github.com/RTKConsortium/RTK/commit/eb90fb42b9dca3fe8be373e2b7c9724e4c3b6156) for an ITK macro. Note the regular expression in the log. Do you want to give it a try or should I try to implement this?

Should we add a target_compile_definition in ITKCudaCommon to peform such a check (e.g. "CudaDataManagerHasGetGPUBufferPointerPtr")? Or do you prefer to use the CudaCommon_VERSION as suggested in https://github.com/RTKConsortium/ITKCudaCommon/pull/42 ? Or something else ?

File: src/rtkCudaWarpForwardProjectionImageFilter.cxx (hunk 1 above)
Comment (LucasGandel) [reply src/rtkCudaWarpForwardProjectionImageFilter.cxx]: Nice catch thanks, it looks like my local build was outdated so I missed that somehow, and even more: it should be a cast to (float *) and not (float **)

**Comment by SimonRit** [comment]
> Should we add a target_compile_definition in ITKCudaCommon to peform such a check (e.g. "CudaDataManagerHasGetGPUBufferPointerPtr")? Or do you prefer to use the CudaCommon_VERSION as suggested in [RTKConsortium/ITKCudaCommon#42](https://github.com/RTKConsortium/ITKCudaCommon/pull/42) ? Or something else ?

No preference, the second one sounds simpler to me. But it is not defined for the latest release so it will just be
```c++
#ifdef CUDA_COMMON_MAJOR_VERSION
#else
#endif
```
I guess

**Comment by LucasGandel** [comment]
@SimonRit To preserve backward compatibility with CudaCommon, it was also required to add checks in itkCudaImageRTK.wrap, as it uses CudaCommon_SOURCE_DIR which is undefined before CudaCommon 2.0.
CudaCommon_SOURCE_DIR is now also used to check if cudaCommonConfiguration.h exists, which will define CUDACOMMON_VERSION_MAJOR.

The linter is unhappy with the new #ifdef, let me know if you have any other feedback before I fix the style.Thanks

**Review by SimonRit — CHANGES_REQUESTED** [review]
This is ready to be merged IMO, thanks for this great work maintaining backward compatibility with the CudaCommon HEAD and the new PR. There are (weird) code style issues:
https://github.com/RTKConsortium/RTK/actions/runs/11798626855/job/32865225878#step:4:6
Can you fix them before I merge please?

**Comment by LucasGandel** [comment]
> This is ready to be merged IMO, thanks for this great work maintaining backward compatibility with the CudaCommon HEAD and the new PR. There are (weird) code style issues: https://github.com/RTKConsortium/RTK/actions/runs/11798626855/job/32865225878#step:4:6 Can you fix them before I merge please?

Thanks! Style issues have been fixed (although they were definitely weird).

**Review by SimonRit — APPROVED** [review]
Thanks for addressing the style issue! BTW, ITK is working on an upgrade of clang, see e.g.
https://github.com/InsightSoftwareConsortium/ITK/pull/5012
https://github.com/InsightSoftwareConsortium/ITK/pull/5015
We should plan following this upgrade.

**Pull request merged**